File: circle_map.go
// Written by http://xojoc.pw. Public domain.

package main

import (
	"image"
	"image/color"
	"log"
	"math"

	"github.com/lucasb-eyer/go-colorful"
	"gitlab.com/xojoc/util"
)

const π = math.Pi

var sin = math.Sin

const width = 300
const height = 300

var canvas = image.NewRGBA(image.Rect(0, 0, width, height))

func θn1(θₙ, Ω, K float64) float64 {
	return θₙ + Ω - (K/(2*π))*sin(2*π*θₙ)
}

func ω(Ω, K float64) float64 {
	θₙ := 0.0
	n := 2000.0
	for i := 0.0; i < n; i++ {
		θₙ = θn1(θₙ, Ω, K)
	}
	return θₙ / n
}

func col(ω float64) color.Color {
	black := colorful.LinearRgb(0, 0, 0)
	green := colorful.LinearRgb(0, 1, 0)
	red := colorful.LinearRgb(1, 0, 0)
	if ω < 1.0/2 {
		return black.BlendLab(green, ω*2).Clamped()
	} else {
		if ω > 1 {
			log.Println("one ", ω)
			ω = 1
		}
		ω = (ω - 0.5) * 2
		if ω > 1 {
			log.Println("2 ", ω)
		}
		return green.BlendLab(red, ω).Clamped()
	}
}

func main() {
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			Ω := float64(x) / width
			K := 2 * π * float64(height-y) / height
			canvas.Set(x, y, col(ω(Ω, K)))
		}
	}
	util.MustWriteImage("circle_map.png", canvas)
}

File: circle_map.py
# Written by http://xojoc.pw. Public domain.
import math
from PIL import Image
import numpy as np
import matplotlib

π = math.pi
sin = math.sin

width = 300
height = 300

#canvas = Image.new("RGB", (width, height))
array = np.zeros((width,height), np.float96)

def circle_map(θ, Ω, K):
    return θ+Ω-(K/(2*π))*sin(2*π*θ)

def ω(Ω, K):
    θ = 0
    n = 1000
    for i in range(n):
        θ = circle_map(θ, Ω, K)
    return θ/n

def color(ω):
    if ω <= 1/2:
        return (0, int(ω*2*255), 0)
    else:
        return (int((ω-1/2)*2*255), int((1-(ω-1/2))*2*255),0)
    
def plot():
    for y in range(height):
        for x in range(width):
            Ω = x/width
            K = y/height*2*π
            #            canvas.putpixel((x, y), color(ω(Ω, K)))
            array[x][y] = ω(Ω, K)

            
colormap = matplotlib.colors.LinearSegmentedColormap.from_list("", [[0.0,0,0], [0,1,0], [1,0,0]])

plot()

Image.fromarray(np.uint8(colormap(array)*255)).save("/tmp/a.png")
